Add naive support for `[ManualAlloc, ChildImpl=virtual]` annotations.

asutherland opened the pull request:
The parser does not seem to like annotations inclusive of both the
`[ManualAlloc]` type added in
https://bugzilla.mozilla.org/show_bug.cgi?id=1736371 and the variations
that also include list of annotations in the patch that look like
`[ManualDealloc, ChildImpl=virtual, ParentImpl=virtual]`.

My efforts here are to try and get this to parse and technically
produce something usable, but I wouldn't be surprised if someone wanted
to properly consume the annotations that they would want something that
looks more strongly typed.  But this is an AST so maybe this is fine?

asutherland commented:
This passes `cargo test` with added test cases and I want to run this against mozilla-central, so I'm going to merge this into the mozsearch branch of ipdl_parser now because of logistical reasons related to git submodules generally being a PITA.  But I think it makes sense for this to be reviewed before making it upstream, so requesting post-landing review.

amccreight commented:
Huh, strange. Looks like I never merged my attribute fixes back to the main branch. I could have sworn I had. I'll update that. Sorry about that. I'll figure out what is going on here with my local changes compared to the main branch. (There's an attributes branch, but it has lots of WIP stuff so don't use that.)

amccreight commented:
Ok, I pushed my changes to https://github.com/amccreight/ipdl_parser now. if you run cargo test in the repo, error_tests will fail, but ok_tests will pass, so it should be okay for parsing the valid stuff in the mozilla-central repo.

asutherland commented:
Fantastic, thanks!

I think maybe I'll move the current mozsearch "master" branch to the side and directly update so we can get back to just fast-forwarding that from your canonical upstream.

asutherland commented:
Okay, yeah, so the current state:
- I moved what had become "master" on mozsearch/ipdl_parser to be the "divergence" branch with all my shenanigans.
- I force-pushed the head state of the "master" branch at https://github.com/amccreight/ipdl_parser to be the "master" branch for the mozsearch fork.  This should get us back to some level of sanity.
- I've updated my https://github.com/mozsearch/mozsearch/pull/528 pull request to:
  - Use this most latest revision.
  - Compensated for the changes in the `Direction` type.
  - Added the very minimal PTestChild "check" tests.
  - `make build-test-repo` now passes.
- I've triggered another "dev" run so we can see how that goes on mozilla-central.  This won't have the extra test logging but I don't think we need it right now and I can spin a new pull request against your upstream that looks better.

asutherland commented:
Note re the force push: I'm treating this repo (https://github.com/mozsearch/ipdl_parser/) as a vendor branch much like we do with livegrep where the intent is that we use it as a submodule and/or manually checked out specific branch from the provisioning scripts, and it's not to be treated as a real tree that people would pull from.

amccreight commented:
Sounds good, thanks.

amccreight commented:
The nice thing about Nika's attributes work is that the actual parser part should require less changes going forward. You can add new attributes and the parser will pass them through without complaint.

asutherland commented:
Nice!
